### PR TITLE
test(pkg/config): fix defaultServerConfig unexpected data race (#2957)

### LINKED ISSUE #2957: defaultServerConfig  in config.go is not thread safe, and will cause data-race.
### What did you do?

When migrating test-infra to testify for ticdc/pkg/config,  I add t.Parallel() to every test-case for the reason that  every test-case new a ServerConfig  object and make some check.   Please ref to [pr 2942](https://github.com/pingcap/ticdc/pull/2942)

### What did you expect to see?

all  test-case can run good in parallel.

### What did you see instead?

go testing check data race fail.

```
==================
WARNING: DATA RACE
Write at 0x0000025f1708 by goroutine 19:
  github.com/pingcap/ticdc/pkg/config.(*ServerConfig).ValidateAndAdjust()
      /Users/maxshuang/test/ticdc/pkg/config/config.go:304 +0x3c4
  github.com/pingcap/ticdc/pkg/config.TestServerConfigValidateAndAdjust()
      /Users/maxshuang/test/ticdc/pkg/config/config_test.go:106 +0x1b0
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1259 +0x22f
  testing.(*T).Run·dwrap·21()
      /usr/local/go/src/testing/testing.go:1306 +0x47

Previous read at 0x0000025f1708 by goroutine 18:
  reflect.Value.String()
      /usr/local/go/src/reflect/value.go:2217 +0x78
  encoding/json.stringEncoder()
      /usr/local/go/src/encoding/json/encode.go:647 +0x1a4
  encoding/json.structEncoder.encode()
      /usr/local/go/src/encoding/json/encode.go:761 +0x2ba
  encoding/json.structEncoder.encode-fm()
      /usr/local/go/src/encoding/json/encode.go:732 +0xdb
  encoding/json.ptrEncoder.encode()
      /usr/local/go/src/encoding/json/encode.go:945 +0x3f1
  encoding/json.ptrEncoder.encode-fm()
      /usr/local/go/src/encoding/json/encode.go:930 +0x90
  encoding/json.structEncoder.encode()
      /usr/local/go/src/encoding/json/encode.go:761 +0x2ba
  encoding/json.structEncoder.encode-fm()
      /usr/local/go/src/encoding/json/encode.go:732 +0xdb
  encoding/json.ptrEncoder.encode()
      /usr/local/go/src/encoding/json/encode.go:945 +0x3f1
  encoding/json.ptrEncoder.encode-fm()
      /usr/local/go/src/encoding/json/encode.go:930 +0x90
  encoding/json.(*encodeState).reflectValue()
      /usr/local/go/src/encoding/json/encode.go:360 +0x88
  encoding/json.(*encodeState).marshal()
      /usr/local/go/src/encoding/json/encode.go:332 +0x22e
  encoding/json.Marshal()
      /usr/local/go/src/encoding/json/encode.go:161 +0x51
  github.com/pingcap/ticdc/pkg/config.(*ServerConfig).Marshal()
      /Users/maxshuang/test/ticdc/pkg/config/config.go:224 +0x37
  github.com/pingcap/ticdc/pkg/config.(*ServerConfig).Clone()
      /Users/maxshuang/test/ticdc/pkg/config/config.go:248 +0x4a
  github.com/pingcap/ticdc/pkg/config.GetDefaultServerConfig()
      /Users/maxshuang/test/ticdc/pkg/config/config.go:347 +0x49
  github.com/pingcap/ticdc/pkg/config.TestServerConfigClone()
      /Users/maxshuang/test/ticdc/pkg/config/config_test.go:89 +0x31
```

Reason:

defaultServerConfig  should be designed as  the singleton and be immutable, but methods like ValidateAndAdjust will write to  defaultServerConfig.Sorter at the object level, which will cause data race unexpected.
```
262 // ValidateAndAdjust validates and adjusts the server configuration                                                                                                             
263 func (c *ServerConfig) ValidateAndAdjust() error {
...
if c.Sorter == nil {
                c.Sorter = defaultServerConfig.Sorter  // shallow copy here, and write below at the object level
}
c.Sorter.SortDir = DefaultSortDir
...


}
```

### Versions of the cluster

Upstream TiDB cluster version (execute `SELECT tidb_version();` in a MySQL client):

```console
5.2.0
```

TiCDC version (execute `cdc version`):

```console
5.2.0
```

### PR DESCRIPTION
**Pull request by CharlesCheung96**
### What problem does this PR solve? 
close #2957 

### What is changed and how it works?
1. Use `GetDefaultServerConfig()` method to get the default configuration instead of directly accessing the `defaultServerConfig` variable
2. Code refactoring: extract `ValidateAndAdjust` method for sorter configuration 

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

Side effects

 - None

Related changes

 - None

### Release note 

```release-note
None
```

### CODE REVIEW

File: pkg/config/config.go [hunk 1]
```diff
@@ -300,8 +300,9 @@ func (c *ServerConfig) ValidateAndAdjust() error {
 
 	if c.Sorter == nil {
 		c.Sorter = defaultServerConfig.Sorter
```
Comment (maxshuang) [inline pkg/config/config.go:302]: I think it's not good to assign defaultServerConfig.Sorter to other point which will be out of control and cause data race.  

something like c.Sorter = GetDefaultServerConfig().Sorter or c.Sorter = defaultServerConfig.Sorter.Clone() is better

Comment (maxshuang) [reply pkg/config/config.go:302]: line 334:  c.KVClient = defaultServerConfig.KVClient 
do other fields in ServerConfig has the same problem?

Comment (maxshuang) [reply pkg/config/config.go:302]: BTW,  pr body need to be more specified

File: pkg/config/config.go [hunk 2]
```diff
@@ -300,8 +300,9 @@ func (c *ServerConfig) ValidateAndAdjust() error {
...
+	} else if c.Sorter != defaultServerConfig.Sorter {
+		c.Sorter.SortDir = DefaultSortDir
```
Comment (3AceShowHand) [inline pkg/config/config.go:304]: for sorter configuration, it's a good time to extract a function to do this.

**Comment by ben1009** [comment]
is the race condition possible for normal case ?

**Review by 3AceShowHand — APPROVED** [review]

**Comment by amyangfei** [comment]
Since this is a bugfix, we should cherry pick to release branches

**Review by amyangfei — APPROVED** [review]

**Pull request merged**